src/core/gosling.schema.ts

```diff
@@ -240,6 +240,19 @@ export interface BrushAndMarkHighlightingStyle {
     opacity: number;
 }
 
+// Mouse events
+type MouseEvent = 'click' | 'mouseOver' | 'rangeSelect';
+export type MouseEventsDeep = {
+    /* Turn on and off individual mouse events. */
+    [Event in MouseEvent]?: boolean;
+} & {
+    /* Group marks using keys in a data field. This affects how a set of marks are highlighted/selected by interaction. __Default__: `undefined` */
+    groupMarksByField?: string;
+
+    /* Determine whether all marks underneath the mouse point should be affected by mouse over. __Default__: `false` */
+    enableMouseOverOnMultipleMarks?: boolean;
+};
+
```

Comment:
This looks good. My only concern is that there is an implicit link between the events/payloads in `api.ts` and`gosling.schema.ts`. The types are not shared or linked in the code at all. 

This means that anywhere we actually publish event data is not covered by our type-system,

```typescript
  PubSub.publish('rangeSelect', "I can put anything here and it's ok!");
```

The type checker does **not** ensure that either the event name _or_ payload are as defined in `api.ts`. Instead, we have to be extremely careful that whatever is in `api.ts` and `PubSub.publish` since no where is it covered by the type system.

One way around this would be to export a utility module that provides type-safe use of `PubSub.{publish,subscribe}`:

```typescript
// pubsub.ts

type EventMap = {
  mouseOver: PointMouseEventData;
  click: PointMouseEventData;
  rangeSelect: RangeMouseEventData;
  rawData: CommonEventData;
}

export function publish<
  Name extends keyof EventMap
>(name: Name, data: EventMap[Name]): void {
  PubSub.publish(name, data);
}

export function subscribe<
  Name extends keyof EventMap
>(name: Name, callback: (msg: string, data: EventMap[Name]) => void): void {
  PubSub.subscribe(name, callback);
}

```

Reply:
Personally, I think it would make sense to define events in `gosling.schema.ts` and then import the `EventMap` in `pubsub.ts/api.ts`:


#### core/gosling.schema.ts

```typescript
export type _EventMap = { /* ... */ } // underscore for internal type

type MouseEvents = {
    [Event in keyof _EventMap]: boolean,

    /* Group marks using keys in a data field. This affects how a set of marks are highlighted/selected by interaction. __Default__: `undefined` */
    groupMarksByField?: string;

    /* Determine whether all marks underneath the mouse point should be affected by mouse over. __Default__: `false` */
    enableMouseOverOnMultipleMarks?: boolean;
}
```

#### core/pubsub.ts

```typescript
import type { _EventMap } from './gosling.schema.ts';

export function publish<
  Name extends keyof _EventMap
>(name: Name, data: EventMap[Name]): void {
  PubSub.publish(name, data);
}

export function subscribe<
  Name extends keyof _EventMap
>(name: Name, callback: (msg: string, data: EventMap[Name]) => void): void {
  PubSub.subscribe(name, callback);
}
```

#### core/api.ts

```typescript
import { subscribe } from './pubsub.ts';

export interface GoslingApi {
    subscribe: typeof subscribe;
    /* ... */
}

export function createApi(
    hg: HiGlassApi,
    hgSpec: HiGlassSpec | undefined,
    theme: Required<CompleteThemeDeep>
): GoslingApi {
    /* .. */
    return {
        subscribe,
        /* ... */
    }
}
```

Reply:
This is a really great idea!

In the [last commit](https://github.com/gosling-lang/gosling.js/pull/712/commits/f23297d3d076e64c705f4533da665e8aa666d860), I updated `publish`, `subscribe`, and `unsubscribe`.